### ISSUE TITLE
Canvas's with multisampling give screwy results.

### ISSUE THREAD

**Issue by icefoxen** [ggez/ggez#253]
See https://postimg.org/gallery/1o6wim2fw/ ; one is with the canvas multisampling = 1, the other is with it = 8.  Changing the screen's multisampling level doesn't seem to affect it.

Generated with this code: https://play.rust-lang.org/?gist=dfc4dbff82d3c8a82f133feadd981b91&version=stable

**Comment by rardiol** [comment]
I think I set the expiration time short, so reuploaded: https://postimg.org/gallery/m4ksyw2k/

**Comment by icefoxen** [comment]
![image](https://user-images.githubusercontent.com/1335133/34834707-a30ab2b8-f6c0-11e7-80c6-79327a8da5d5.png)
![image](https://user-images.githubusercontent.com/1335133/34834709-a6aa1be8-f6c0-11e7-9bd8-deb51f448e06.png)

**Comment by icefoxen** [comment]
This is very vexing because now that I finally get around to looking at it more deeply, I can't reproduce it on my Intel graphics laptop.  Try this on other computers!

Test program (just noodle with the `NumSamples` setting):

```rust
//! Basic hello world example.

extern crate ggez;

use ggez::conf;
use ggez::event;
use ggez::{Context, GameResult};
use ggez::graphics;
use std::env;
use std::path;

struct MainState {
    text: graphics::Text,
    canvas: graphics::Canvas,
    frames: usize,
    draw_with_canvas: bool,
}

impl MainState {
    fn new(ctx: &mut Context) -> GameResult<MainState> {
        // The ttf file will be in your resources directory. Later, we
        // will mount that directory so we can omit it in the path here.
        let font = graphics::Font::new(ctx, "/DejaVuSerif.ttf", 48)?;
        let text = graphics::Text::new(ctx, "Hello world!", &font)?;
        let canvas = graphics::Canvas::with_window_size(ctx)?;

        let s = MainState { text, canvas, draw_with_canvas: false, frames: 0 };
        Ok(s)
    }
}

impl event::EventHandler for MainState {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::set_background_color(ctx, graphics::Color::from((255, 0, 0, 0)));
        graphics::clear(ctx);
        if self.draw_with_canvas {
            graphics::set_canvas(ctx, Some(&self.canvas));
        }
        graphics::set_background_color(ctx, graphics::Color::from((0, 0, 0, 255)));
        graphics::clear(ctx);

        // Drawables are drawn from their top-left corner.
        let dest_point = graphics::Point2::new(10.0, 10.0);
        graphics::draw_ex(ctx, &self.text, graphics::DrawParam {
            dest: dest_point, 
            color: Some(graphics::Color::from((128, 128, 128, 255))),
            .. Default::default()
        } )?;
        graphics::set_canvas(ctx, None);
        graphics::draw(ctx, &self.canvas, graphics::Point2::new(0.0, 0.0), 0.0)?;
        graphics::present(ctx);

        self.frames += 1;
        if (self.frames % 100) == 0 {
            println!("FPS: {}", ggez::timer::get_fps(ctx));
        }

        Ok(())
    }

    fn key_down_event(
        &mut self, 
        _ctx: &mut Context, 
        _keycode: ggez::event::Keycode, 
        _keymod: ggez::event::Mod, 
        repeat: bool) {
            if !repeat {
                self.draw_with_canvas = !self.draw_with_canvas;
                println!("Canvas on: {}", self.draw_with_canvas);
            }
    }
}

pub fn main() {
    let c = conf::Conf {
        window_setup: conf::WindowSetup {
            samples: conf::NumSamples::Eight,
            ..Default::default()
        },
        .. Default::default()
    };
    let ctx = &mut Context::load_from_conf("helloworld", "ggez", c).unwrap();

    if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
        let mut path = path::PathBuf::from(manifest_dir);
        path.push("resources");
        ctx.filesystem.mount(&path, true);
    }

    let state = &mut MainState::new(ctx).unwrap();
    if let Err(e) = event::run(ctx, state) {
        println!("Error encountered: {}", e);
    } else {
        println!("Game exited cleanly.");
    }
}

```

**Comment by icefoxen** [comment]
Fixing the test program reproduced the error:

![image](https://user-images.githubusercontent.com/1335133/40585304-2bc3a038-617e-11e8-85b2-8aaf3545298b.png)

**Comment by icefoxen** [comment]
Okay, observations:

 * Setting the alpha of the text to 1 still makes those weird fringes appear
 * This happens with any image, not just text

![image](https://user-images.githubusercontent.com/1335133/40585796-7a80376a-6186-11e8-9270-df85f2b8b951.png)

**Comment by icefoxen** [comment]
You have to set the background to a color that is not white or black for it to show up, which is why we've never seen this before.

Turning off sRGB conversion in shader stuff doesn't seem to make a difference.

The weird raster-y little chunk-y color fringes don't seem to change size with multisampling level.

**Comment by icefoxen** [comment]
Progress, now that I'm awake!  To duplicate, you should be able to just run the `hello_canvas` example added in b3fbc9e9a9a3bb0285af27bfdc5b86d5fa2b765d and change the number of samples in the `Conf` setup from one to anything else.

**This only happens on my laptop with Intel graphics**, using Debian testing.  On my Windows+AMD system and my Linux+NVidia systems, this does not happen!

**Comment by icefoxen** [comment]
Seems to have fixed itself: #523
